File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.ts
function lengthOfLongestSubstring(S: string): number {
  if(!S || S.length == 0)
    return 0;
  
  const N = S.length,
        map = new Map<string, number>();
  let ans = 0;
  
  for(let i = 0, j = 0; i < N; i++) {
    
    if(map.has(S.at(i))) {
      j = Math.max(j, 1 + map.get(S.at(i)))
    }

    map.set(S.at(i), i);
    ans = Math.max(ans, i - j + 1);
  }
  
  return ans;
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String S) {
        int ans = 0, N = S.length();
        Set<Character> seen = new HashSet<>();
        
        for(int left = 0, right = 0; right < N; ++right) {
            while(seen.contains(S.charAt(right))) {
                ans = Math.max(ans, seen.size());
                seen.remove(S.charAt(left++));
            }

            seen.add(S.charAt(right));
        }
        
        return Math.max(ans, seen.size());
    }
}